composer/callbacks/generate.py

```diff
@@ -0,0 +1,116 @@
+            :attr:`.TimeUnit.TOKEN`, or :attr:`.TimeUnit.SAMPLE`.
+        batch_size (int): Size of a prompt batch for generation.
+        kwargs: All kwargs well be passed along to the call to generate. This is for things like `do_sample`, `top_p`, etc
+    """
+
+    def __init__(self, prompts: List[str], interval: Union[str, int, Time], batch_size: int, **kwargs: Any):
+        self.prompts = prompts
+        self.generate_kwargs = kwargs
+        self.batch_size = batch_size
+        self.check_interval = create_interval_scheduler(interval, include_end_of_training=False)
+
+    def run_event(self, event: Event, state: State, logger: Logger) -> None:
+        if state.get_elapsed_duration() is not None and self.check_interval(state, event):
+            self.generate(state, logger)
+        else:
+            super().run_event(event, state, logger)
+
+    def generate(self, state: State, logger: Logger):
+        model = state.model
+        if isinstance(model, DistributedDataParallel):
+            # Why does DistributedDataParallel wrap the HuggingFaceModel and not the other way around?
+            # Can we make this assumption?
+            model = model.module
+        assert isinstance(
+            model, HuggingFaceModel
+        ), f'Expected HuggingFaceModel, but got {state.model.__class__}'  # TODO: Extend to support any models that have a generate method.
+
+        tokenizer = model.tokenizer
+
+        try:
+            from transformers import PreTrainedTokenizer, PreTrainedTokenizerFast
+        except ImportError as e:
+            raise MissingConditionalImportError(extra_deps_group='nlp',
+                                                conda_package='transformers',
+                                                conda_channel='conda-forge') from e
+        Tokenizer = Union[PreTrainedTokenizer, PreTrainedTokenizerFast]
+        tokenizer = cast(Tokenizer, tokenizer)
+
+        # Set to evaluation mode and stash the original mode.
+        original_mode = model.training
+        model.eval()
+        device = state.device
+
+        # Stash the original value of padding_side because generation requires left padding
+        original_padding_side = tokenizer.padding_side
+        tokenizer.padding_side = 'left'
+        if tokenizer.pad_token_id is None:
+            tokenizer.pad_token_id = tokenizer.eos_token_id
+        tokenized_input = tokenizer(self.prompts, return_tensors='pt', padding=True)
+
+        for k, v in tokenized_input.items():
+            tokenized_input[k] = device.tensor_to_device(v)
+
+        all_input_ids = tokenized_input['input_ids']
+        all_attn_masks = tokenized_input['attention_mask']
+
+        batches = DataLoader(TensorDataset(all_input_ids, all_attn_masks), self.batch_size)
+
+        output_token_ids = []
+        # dummy forward call needed for FSDP to work consistently
+        model.dummy_forward_called = False  #
```

Comment:
This is necessary because the code in model.generate sets `dummy_forward_called` to True after the first call to `generate`, which causes a cuda memory error due to the dummy forward pass not being called in subsequent train batches. Perhaps we should be modifying the `HuggingFaceModel.generate` function to address this? @dakinggg

Reply:
Do you think it make sense to modify `HuggingFaceModel` to set `dummy_forward_called` to `False` after a forward call with gradients enabled? (like this: https://github.com/mosaicml/composer/commit/a3b12c2e02a5537cfcd4327897fbf2d23cfd976a#diff-e61f9af34de66238d99995ebdc5d2645cea3c7ca7c303ac006239a720c5b620cR361) That seems to work...

I _think_ this makes sense because the dummy would be needed to help free the parameters after a (training) forward call

Reply:
If a proper fix is more involved than this, it makes sense to not bother fixing bc of torch2 fix 😛

Reply:
I vote to just not fix it in `HuggingFaceModel.generate`. The solution wasn't that robust to start with, and I'm not sure how robust that line you added is (although it seems totally reasonable).

Reply:
sgtm!